chore: add additional information around Cypress configuration (#4202)
cypress-io/cypress-documentation: content/api/cypress-api/config.md

### Not all config values can be changed at all times

Some configuration values are readonly and cannot be changed while running a test. Anything
that's not directly under Cypress's control - like timeouts, `userAgent`, or
environment variables - will be ignored at run-time. Be sure to review the list of
[test-time configuration options](/guides/references/configuration##Test-time-overrides)

> ## Test Configuration vs `Cypress.config()`

Reviewer: Changing this heading may have broken other links in the docs - need to look for anywhere that might be references `content/api/cypress-api/config#Test-Configuration` and fix all of those links.

Author: There are no links to `content/api/cypress-api/config#Test-Configuration`

suite or test where they are set. The values will then reset to the previous
default values after the suite or test is complete.

> See the full guide on [test-time configuration](/api/cypress-api/config#Test-time-overrides).

Reviewer: I don't see this link, where is this?

## `Cypress.config()` executes Synchronously

It's important to note that `Cypress.config()` executes synchronously and will not wait for the Cypress commands above it to execute.

> If you need to update your configuration mid-test, be sure to either chain the command the
> [asynchronously Cypress command](/guides/core-concepts/introduction-to-cypress#Commands-Are-Asynchronous) before it or wrap execute it with
> the `cy.then` Cypress command to correctly queue.

Reviewer: We don't recommend chaining `.then` directly off of `cy`. If we decide to enforce chaining in a later version this would cause a lot of issues for people we recommended it to.

```suggestion
If you need to update your configuration mid-test, be sure to either chain the command the
[asynchronously using a Cypress command](/guides/core-concepts/introduction-to-cypress#Commands-Are-Asynchronous) before it.
```

```javascript
it('using cy.then', () => {
  cy.visit('/my-test_page')

>   cy.click('#download-html')
>   cy.then(() => {
>     Cypress.config('baseUrl', 'null')
>   })

Reviewer: ```suggestion
  cy.click('#download-html')
    .then(() => {
      Cypress.config('baseUrl', 'null')
    })
```
